src/main/java/com/google/devtools/build/lib/remote/RemoteActionInputFetcher.java

```diff
@@ -158,6 +159,12 @@ void downloadFile(Path path, FileArtifactValue metadata)
     }
   }
 
+  private static Path toTmpDownloadPath(Path actualPath) {
+    return actualPath.getParentDirectory().getRelative(
+      actualPath.getBaseName() + "."  + actualPath.hashCode()
```

Comment:
This should give a unique + deterministic enough name so that we don't need to [sort all downloads before doing this](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/remote/RemoteCache.java;l=406;drc=f17c403b2365ff0268b0ecae2e00e95fcad9d36b?q=RemoteCache&ss=bazel%2Fbazel).

Reply:
Why not use the same approach as RemoteCache?

Reply:
Because RemoteCache collects all outputs before doing the [moveToFinalLocation](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/remote/RemoteCache.java;l=372?q=RemoteCache&ss=bazel). This class was designed a bit differently in that it can call downloadFileAsync from two different methods ([downloadFile](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/remote/RemoteActionInputFetcher.java;l=149?q=RemoteActionInputFetcher&ss=bazel) and [prefetchFiles](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/remote/RemoteActionInputFetcher.java;l=89?q=RemoteActionInputFetcher&ss=bazel)) and the shared variables `downloadsInProgress` and `downloadedPaths` are controlled per download here.

So I would have to move the `synchronized` blocks to those two methods and that looked like a lot more refactoring to me. I can make that change if you think that is the way to go. But this seems cleaner to me.